xdp/src/device.rs

```diff
@@ -0,0 +1,495 @@
+    std::{
+        ffi::{c_char, CStr, CString},
+        io::{self, ErrorKind},
+        marker::PhantomData,
+        mem,
+        net::Ipv4Addr,
+        os::fd::{AsRawFd as _, FromRawFd as _, OwnedFd, RawFd},
+        ptr, slice,
+        sync::atomic::{AtomicU32, Ordering},
+    },
+};
+
+#[derive(Copy, Clone, Debug)]
+pub struct QueueId(pub u64);
+
+pub struct NetworkDevice {
+    if_index: u32,
+    if_name: String,
+}
+
+impl NetworkDevice {
+    pub fn new(name: impl Into<String>) -> Result<Self, io::Error> {
+        let if_name = name.into();
+        let if_name_c = CString::new(if_name.as_bytes())
+            .map_err(|_| io::Error::new(ErrorKind::InvalidInput, "Invalid interface name"))?;
+
+        let if_index = unsafe { libc::if_nametoindex(if_name_c.as_ptr()) };
+
+        if if_index == 0 {
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(Self { if_index, if_name })
+    }
+
+    pub fn new_from_index(if_index: u32) -> Result<Self, io::Error> {
+        let mut buf = [0u8; 1024];
+        let ret = unsafe { libc::if_indextoname(if_index, buf.as_mut_ptr() as *mut c_char) };
+        if ret.is_null() {
+            return Err(io::Error::last_os_error());
+        }
+
+        let cstr = unsafe { CStr::from_ptr(ret) };
+        let if_name = String::from_utf8_lossy(cstr.to_bytes()).to_string();
+
+        Ok(Self { if_index, if_name })
+    }
+
+    pub fn new_from_default_route() -> Result<Self, io::Error> {
+        let router = Router::new()?;
+        let default_route = router.default().unwrap();
+        NetworkDevice::new_from_index(default_route.if_index)
+    }
+
+    pub fn name(&self) -> &str {
+        &self.if_name
+    }
+
+    pub fn if_index(&self) -> u32 {
+        self.if_index
+    }
+
+    pub fn mac_addr(&self) -> Result<[u8; 6], io::Error> {
+        let fd = unsafe { libc::socket(libc::AF_INET, libc::SOCK_DGRAM, 0) };
+        if fd < 0 {
+            return Err(io::Error::last_os_error());
+        }
+        let fd = unsafe { OwnedFd::from_raw_fd(fd) };
+
+        let mut req: ifreq = unsafe { mem::zeroed() };
+        let if_name = CString::new(self.if_name.as_bytes()).unwrap();
+
+        let if_name_bytes = if_name.as_bytes_with_nul();
+        let len = std::cmp::min(if_name_bytes.len(), IF_NAMESIZE);
+        unsafe {
+            std::ptr::copy_nonoverlapping(
+                if_name_bytes.as_ptr() as *const c_char,
+                req.ifr_name.as_mut_ptr(),
+                len,
+            );
+        }
+
+        let result = unsafe { ioctl(fd.as_raw_fd(), SIOCGIFHWADDR, &mut req) };
+        if result < 0 {
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(unsafe {
+            slice::from_raw_parts(req.ifr_ifru.ifru_hwaddr.sa_data.as_ptr() as *const u8, 6)
+        }
+        .try_into()
+        .unwrap())
+    }
+
+    pub fn ipv4_addr(&self) -> Result<Ipv4Addr, io::Error> {
```

Comment:
If interface has multiple addresses, this returns first, right? Should we maybe consider pulling all addresses and returning a vec<ipv4addr> instead?

Reply:
It returns the "primary" address. 

I don't mind making it return a vec an then picking the first in the caller, but that effectively has exactly the same behavior, and dealing with `getifaddrs` is slightly more annoying because it returns all the addresses for all interfaces.

Reply:
My concern is only that we could get annoying issue reports:) I guess this is fine as is for now.